replace coloring code with coloring plugin

lucky01 opened the pull request:
Hi Freezy,

as recommended from you here is the pull request with the code changes necessary to make the pin2dmd coloring plugin with pac support work. Also some fixes for pindmdV3 and alphanumeric devices integrated.

lucky01 commented:
@mista-funky

lucky01 commented:
@nailbuster 
I made this change to the code to avoid a lag of videos and sounds beeing played due to retriggering
https://github.com/freezy/dmd-extensions/pull/347/commits/c15f3dc2d9b3e43266dfd5a984f5ae68c72543de
Could this cause triggering problems that the frame only triggers once ?

freezy commented:
Thanks!

I'll need some time for this. Hopefully next week.

lucky01 commented:
Take your time ! No rush. My testing release of this code should be really stable now according to the reports of the community.

nailbuster commented:
@lucky01 not sure, what do you mean by "frame only triggers once"?  How does it differ from the way frames were sent in prior versions?

Id guess testing will know for sure.  Would this frame-change only be while using color-roms with pups, or does this change for all types?

lucky01 commented:
> @lucky01 not sure, what do you mean by "frame only triggers once"? How does it differ from the way frames were sent in prior versions?
>
As you can see from the code change your dll only gets the frame sent when it has changed, while in the old version it get the frame with the same content sent periodically in the frequency of pinmame frame updates. 
 
> Id guess testing will know for sure. Would this frame-change only be while using color-roms with pups, or does this change for all types?
It changes for all types. The change was introduced because of users reporting a lag of callouts and videos which seems to be caused by the same trigger beeing hit multiple times due to the periodic update. 
See here https://vpuniverse.com/forums/topic/7504-new-version-dmd-extensions-with-pin2dmd-coloring-plugin-and-pac-file-support/?do=findComment&comment=64588
The problem is that after that change some pup packs don´t work properly because of triggers beeing missed from pup player.

lucky01 commented:
@nailbuster Maybe you can stop the retriggering lag on your side ?

nailbuster commented:
@lucky01 Not trying to be difficult, but there's thousands of users and hundreds of puppacks and also pup-authoring tools with delays/ms timing setup within many packs.   

I don't think its wise to break backwards compatibility (if that is what will happen?) with all these users/pups/setups.  Can we not send frames to the pup driver exactly like before, without having users update pup files?

lucky01 commented:
@nailbuster I don´t break backwards compatibility because I already reverted back to old code as it did not work out with other pup packs. But that way we have the trigger lag problem back which was fixed by my code change. That´s why I asked you wether you can handle that issue on your side. That way we can keep sending the frames in the old format AND have the lag removed.

lucky01 commented:
@nailbuster Although it makes me wonder why I have to forward the same frame multiple times to your pup dll . Could it be that you miss triggers otherwise ? At least that is what members of the community are experiencing.

nailbuster commented:
@lucky01 I haven't looked at the dmdext code in years, so all I can suggest is that it works great with the old method in sending frames to pupdll.  Don't know what has/is being changed that would make users report getting lag?   Also, best to make sure we aren't chasing other issues.

Maybe threading of sending frames has changed?  

IIRC,  pupdll only use to get frame updates when contents 'changed'... was not getting constant frames updates( like 30fps or 60fps updates).    

Also,  if frames are being missed,  maybe it has to do with different scaling or dot sizes or something that is modifying frames before being sent to pupdll...before pupdll would get 'virgin' frames from vpinmame i assume.  Any scale/hd/dot sizes maybe are altering the frames prior to pupdll?

lucky01 commented:
@nailbuster It has nothing to do with scaling etc. The frames are sent like they come from pinmame.
My version of the plugin code without the code change to the pup output seems to work exactly like current freezy release and no frame triggers are missed that way.
The only problem is the lag which is discussed here https://vpuniverse.com/forums/topic/7504-new-version-dmd-extensions-with-pin2dmd-coloring-plugin-and-pac-file-support/?do=findComment&comment=64585 , which was fixed by my code change, is back again. I´m pretty sure that it would also be there if the change would be applied to the original freezy.
Is there any chance you can fix the lag on your side ? I assume it is caused by the same trigger beeing hit several times in a row.

freezy commented:
So I've finally started looking at this, thanks again for getting this back upstream. I'm not completely through yet, but here are my first comments:

- In order to keep backwards-compatibility, I'm in favor of keeping the current coloring code intact, so people who don't want to upgrade to the new coloring format don't have to (also, there are issues with PUP, it seems).
- As far as I can see, the new plugin API only supports RGB24 frames in return. This is sub optimal, since other displays aren't able to leverage the more optimized frame formats. I don't see why the most optimal format can be returned and converted by dmdext to whatever the active output device(s) support.
- It would be nice to move all the boilerplate code for the plugin loading to a separate class instead of `DmdDevice.cs`.
- I'm not a fan of hard-coding external dependencies. I would rather prefer a configuration in `DmdDevice.ini` that points to any plugin(s) that should be loaded. This would add flexibility and remove references to (currently) proprietary software that I'd like to avoid. I can help with this if needed.

Given we now have two (actually, three with legacy) coloring solutions, I would love to abstract this through an API. But that's my own department, so I'll have a look at that when the rest is done.

@lucky01 I would propose to rebase your changes on top of #349 while taking into account my comments. I can do it myself as well, but then we'd lose your authorship of the changes in git, which would be a pity.

lucky01 commented:
> In order to keep backwards-compatibility, I'm in favor of keeping the current coloring code intact, so people who don't want to upgrade to the new coloring format don't have to (also, there are issues with PUP, it seems).

If you want you can revert the commits which removed my coloring code but that will lead to further support issues since I added two new coloring modes "Replace No Retrigger" and "ColorMask Loop" for the upcoming alphanumeric colorizations. So even if the author exports in pal/vni format, they will not work. Another thing which was very disrespectful against the colorization authors, was the sharing of vni/pal files in a pack in the installation instructions for ZeDMD without their approval.
You know how sensitive the scene is with illegal sharing of their work. The issue with pup was caused by a code change I made to the pup driver. I already reverted that code change and now it is working exactly like official dmdext.

> As far as I can see, the new plugin API only supports RGB24 frames in return. 

The only devices natively supporting Colored6Gray frames are PIN2DMD and ZeDMD, all others have to convert to RGB24 anyway. For pin2dmd I have native USB drivers in the coloring plugin which get activated if you set pin2dmd enabled to native instead of true in the dmddevice.ini. The other will not be supported by my coloring plugin anyways. I would slap the pin2dmd coloring authors in the face if I would do that. There is already enough beef going on. (See above)

> It would be nice to move all the boilerplate code for loading the plugin into a separate class instead of DmdDevice.cs.

I´m open for code suggestions.

> I'm not a fan of hard-coding external dependencies. I would rather prefer a configuration in DmdDevice.ini that points to any plugin(s) that should be loaded. This would add flexibility and remove references to (currently) proprietary software that I'd like to avoid. I can help with this if needed.

I´m open for code suggestions, but dmddevicepup.dll is also hardcoded and making it configurable is more difficult for the end user instead of just copying next to dmddevice.dll

zesinger commented:
> Another thing which was very disrespectful against the colorization authors, was the sharing of vni/pal files in a pack in the installation instructions for ZeDMD without their approval. You know how sensitive the scene is with illegal sharing of their work.

You are perfectly right, it was a mistake I am 100% responsible of and would like to apologize for that! Now the links have been removed and the files have been deleted!

freezy commented:
> So even if the author exports in pal/vni format

I was referring to existing setups that are compatible today. I actually assumed that you've removed VNI export from PIN2DMD editor, but even if it's still there, I guess that the coloring authors all changed to PAC anyway.

> You know how sensitive the scene is with illegal sharing of their work. 

Yes, but I don't see the relevance to breaking backwards compatibility. Why do you mention that?

> The other will not be supported by my coloring plugin anyways. I would slap the pin2dmd coloring authors in the face if I would do that.

Oh, so you meant the coloring plugin to be compatible only with PIN2DMD and the virtual DMD? Because of... revenge?

Look, dmdext won't take any side of whatever beef you have with whoever. So I won't participate in a system that privileges users of hardware X over users with hardware Y. You know that @zesinger created a new coloring system. With dmdext, PIN2DMD users will equally profit off colorizations done for that. I'm working on Scorbit support for dmdext. PIN2DMD users will equally profit from that. But I can't do that with RGB24 frames.

> I´m open for code suggestions, but dmddevicepup.dll is also hardcoded and making it configurable is more difficult for the end user instead of just copying next to dmddevice.dll

Then that's a good opportunity to change that too. And I can also *not* use a different default path to what users are using today, and not break their setup or make it more difficult. ;)

lucky01 commented:
> You are perfectly right, it was a mistake I am 100% responsible of and would like to apologize for that! Now the links have been removed and the files have been deleted!

I know and I personally appreciate this, but that does not change the mind of the colorization authors immediatly. I guess it will take some time.

lucky01 commented:
> I was referring to existing setups that are compatible today. I actually assumed that you've removed VNI export from PIN2DMD editor, but even if it's still there, I guess that the coloring authors all changed to PAC anyway.

All the devices that were in the 10.2. release are supported by the coloring plugin.  All unencrypted file formats are still integrated in the editor. I leave the choice to the colorization authors which format to use and it seems most of them prefer
the encrypted pac format because it prevents unwanted use of their work. One of the unwanted use is the use with ZeDMD because of David sharing a colorization pack without the approval of the authors. If I would allow the coloring plugin and pac to be used with ZeDMD I would act against the will of the colorization authors.
> 
> > You know how sensitive the scene is with illegal sharing of their work.
> 
Because that is what David did and some of the authors therefore don´t want their work to be used with ZeDMD.

> Oh, so you meant the coloring plugin to be compatible only with PIN2DMD and the virtual DMD? Because of... revenge?
No with all devices up to 10.2. . Not because of  revenge but to respect the will of the authors.
> 
> Look, dmdext won't take any side of whatever beef you have with whoever. So I won't participate in a system that privileges users of hardware X over users with hardware Y. You know that @zesinger created a new coloring system. With dmdext, PIN2DMD users will equally profit off colorizations done for that. I'm working on Scorbit support for dmdext. PIN2DMD users will equally profit from that. But I can't do that with RGB24 frames.

And why is that ? The original uncolorized frames are still available for Scorbit. Just like for PUP Player.
I totally appreciate the effort David puts into the new coloring system, but from my side I have to respect the decision of the colorization authors and I will definitly not act against them. Of course you can bring back the old colorization code for backwards compatibility and if the colorization authors avoid the new colorization modes the VNI export should still work like before, with all the little glitches I was not able to fix.
> 
> > I´m open for code suggestions, but dmddevicepup.dll is also hardcoded and making it configurable is more difficult for the end user instead of just copying next to dmddevice.dll
> 
> Then that's a good opportunity to change that too. And I can also _not_ use a different default path to what users are using today, and not break their setup or make it more difficult. ;)

freezy commented:
> One of the unwanted use is the use with ZeDMD because of David sharing a colorization pack without the approval of the authors.

That's pretty much the definition of revenge, no? "I don't let you to do this thing because you did that other thing."

freezy commented:
Maybe, instead of pushing for a position I doubt *all* the coloring authors stand behind, maybe you could also explain to them that it was a mistake and get them to relax their position so we can find a solution where everybody benefits?

lucky01 commented:
> maybe you could also explain to them that it was a mistake, that David is sorry, and get them to relax their position so we can find a solution where everybody benefits?

Sorry to say but I don´t think this is my job, aside from the fact that I don´t have the time to do that.

freezy commented:
> Sorry to say but I don´t think this is my job, aside from the fact that I don´t have the time to do that.

This is an unusual answer from you - are you alright? ;)

lucky01 commented:
> This is an unusual answer from you - are you alright? ;)

The older I get the more I choose my battles.

zesinger commented:
> > You are perfectly right, it was a mistake I am 100% responsible of and would like to apologize for that! Now the links have been removed and the files have been deleted!
> 
> I know and I personally appreciate this, but that does not change the mind of the colorization authors immediatly. I guess it will take some time.

Sure, that's normal!

zesinger commented:
Please, I am the only one responsible for that and don't want others to suffer the consequences. For the authors, I am sorry but can't do anything more than apologize.
@freezy , thank you for the support.

lucky01 commented:
@zesinger You don´t have to apologize to me. I´m not a coloring author. I simply need the approval of every single colorrization author that he is o.k that pac is supported on your device.

freezy commented:
> The older I get the more I choose my battles.

How about this: Instead of creating a battle, get the authors together and we work something out. I'm sure there is a way they can accept David's apology. I agree it was a stupid move to publish those files, but David is a contributor to this hobby as well, and it's not like he's done it because of greed or any other nasty motive. It was a mistake, and I believe you guys should be able to pardon him in some way, so we can move forward and find a solution where the majority of the community is happy.

We can do it in a PM at VPU, you would just need to create the group, or PM me the usernames so Dazz can create a private club or board or whatever.

lucky01 commented:
I think you misunderstood my sentence. Choose your battles means to be selective of the problems, arguments, and confrontations that you get involved in. Instead of fighting every problem, you save your time only for the things that matter. This means fighting the most important battles and letting go of the rest.

I don´t keep track of the colorizations and their authors but here is a list https://vpuniverse.com/forums/topic/5391-complete-pin2dmd-colour-file-list-for-virtual-real-pins/

lucky01 commented:
... and please do me a favour don´t bash any of the authors if he gives a answer you don´t want to hear !  It is everybodys own decision what to do with his work and we have to respect that.

Phil1972 commented:
Any news on when this PR will be merged or is there on ongoing debate?  Thanks

lucky01 commented:
> Any news on when this PR will be merged or is there an ongoing debate? Thanks

Unitl it is merged you can test it with the version from here https://vpuniverse.com/forums/topic/7504-new-version-dmd-extensions-with-pin2dmd-coloring-plugin-and-pac-file-support/

Phil1972 commented:
Thanks for the quick response Lucky, I already tested that version and its fine but I am adding some new functionality for the ease of DMD placement and I would have liked to have your code into the master branch.  But its ok no rush I can manage for now its just that it would be cleaner to use the master branch rather than the way it is now.  thanks,

lucky01 commented:
Updated to the latest codebase of the master branch, fixed some leftover issues and according to the community feedback is working perfecly stable.

lucky01 commented:
@freezy Of course I´m o.k. with the solution of coverting the provided RGB24 data to coloredgray6 on the fly.  When you say you pull my plugin solution into your master branch do you also plan to remove the old pal/vni support ? I think it would be cleaner to do this since newer releases use coloring modes like "replace no retrigger" which don´t work with that implementation and would lead to new support issues.

freezy commented:
So, I've started looking at this. I copied and renamed the Congo colorization to `VPM/altcolor/pin2dmd.pac`. Then I started VPM's setup.exe with this branch.

- `Pin2Color.Open()` returns `false`. Does this mean it didn't find the PAC file?
- Is there a way you could add a function where I can pass a log function to your DLL? That would make debugging much easier, as right now, it's a black box. Also, a function that returns the version of `pin2color.dll` would be nice, so we can log it.

lucky01 commented:
I sent you a dll version for testing. I refactored the coloring plugin interface to use the BSD-3 licensed dmddevice.dll api. 
With that it should be possible to embed the pin2color dll in your releases if your want.
I would like to revoke the GPL-V2 license granted to read the vni/pal file structure with the current libdmd/converter/colorize code to avoid enforcement of vni/pal colorizations to be illegally shared. Especially since the new FX3 / pinballarcade color function and the possible use of the dmddevice.dll api for FX. Please don´t don´t use that code in future releases of dmdext.
Thanks !

freezy commented:
> I sent you a dll version for testing. I refactored the coloring plugin interface to use the BSD-3 licensed dmddevice.dll api.
With that it should be possible to embed the pin2color dll in your releases if your want.

Cool, but no, that still doesn't allow anyone to bundle pin2color.dll with dmdext, unless pin2color.dll itself comes with a GPL2-compatible license.

> I would like to revoke the GPL-V2 license granted to read the vni/pal file structure with the current libdmd/converter/colorize code to avoid enforcement of vni/pal colorizations to be illegally shared.

Again, that's not possible and neither desirable, because it would break users' current setup.

freezy commented:
So, with your latest DLL and this branch, I still got this:

```
 [1] 2023/05/06 11:44:33.284  INFO | [Pin2Color] Successfully loaded colorizer plugin ... 
 [1] 2023/05/06 11:44:33.290  INFO | [Pin2Color] Looking for colorization at L:\Visual Pinball\VPinMAME\altcolor\congo_21 ... 
 [1] 2023/05/06 11:44:33.290  INFO | [Pin2Color] No colorization found. Switching tp Passthrough. 
 [1] 2023/05/06 11:44:33.290  INFO | [Pin2Color] Colorizer v4.50.001 initialized. 
 ```

```
 Directory of L:\Visual Pinball\VPinMAME\altcolor\congo_21

06/05/2023  11:47    <DIR>          .
22/04/2023  00:25    <DIR>          ..
02/05/2023  23:14         1’517’217 pin2dmd.pac
18/03/2020  00:01            16’865 pin2dmd.pal
18/03/2020  00:01         8’998’851 pin2dmd.vni
               3 File(s)     10’532’933 bytes
```

Maybe a logging function inside the DLL would be better, instead of assuming it just works and returns the right values?

lucky01 commented:
Two colorization formats in one folder ? Did you try remove/rename the pac ?

freezy commented:
Same result.
```
 [1] 2023/05/06 17:15:41.613  INFO | [Pin2Color] Successfully loaded colorizer plugin ... 
 [1] 2023/05/06 17:15:41.618  INFO | [Pin2Color] Looking for colorization at L:\Visual Pinball\VPinMAME\altcolor\congo_21 ... 
 [1] 2023/05/06 17:15:41.618  INFO | [Pin2Color] No colorization found. Switching tp Passthrough. 
 [1] 2023/05/06 17:15:41.618  INFO | [Pin2Color] Colorizer v4.50.001 initialized. 
```
```
 Directory of L:\Visual Pinball\VPinMAME\altcolor\congo_21

06/05/2023  17:15    <DIR>          .
06/05/2023  16:04    <DIR>          ..
18/03/2020  00:01            16’865 pin2dmd.pal
18/03/2020  00:01         8’998’851 pin2dmd.vni
02/05/2023  23:14         1’517’217 _pin2dmd.pac
               3 File(s)     10’532’933 bytes
```

lucky01 commented:
It still has the pac extension, so it is still a pac file. rename to e.g pin2dmd.bkp

lucky01 commented:
> Again, that's not possible and neither desirable, because it would break users' current setup.
They are still allowed to use there current setup up to the current version of your dll, but for every new version they would simply have to install the plugin to get the full vni/pal functionality plus back plus extras.

I think we both know that wether revokation in that case is possible or not could only be decided at court and not with you and me stating our positions over and over again. For my side e.g. I found this statement suitable https://news.ycombinator.com/item?id=19827448

Since I´m the original licensor of the key components of the colorization code I hereby revoke the licence to use any of my contributions in that part of your project. 

That said it is on you now wether you want to continue with a possible license violation from your side by keeping my code active or maybe we find a solution we can both live with for the future.  On thing is clear, having a code active which already is outdated from the functional side could lead to more user unfriendly environment than removing VNI/PAL support and pointing to my DLL.

freezy commented:
> Since I´m the original licensor of the key components of the colorization code I hereby revoke the licence to use any of my contributions in that part of your project.

That's absurd, and you're making a fool out of yourself. 

At least be consistent with your argumentation. Above, you talked about the ability to "read the vni/pal file structure", now you're referring to the actual source code you've contributed.

> On thing is clear, having a code active which already is outdated from the functional side could lead to more user unfriendly environment than removing VNI/PAL support and pointing to my DLL.

No, and that's why: Users who are already running with dmdext+VNI/PAL don't care the coloring is "outdated". If they would, they would already have switched to your DLL. Removing VNI/PAL from dmdext would break their setup. And users who are running your fork aren't impacted anyway.

lucky01 commented:
Do you think insulting me does help ? Am 06.05.2023 um 18:40 schrieb freezy ***@***.***>:﻿

Since I´m the original licensor of the key components of the colorization code I hereby revoke the licence to use any of my contributions in that part of your project.

That's absurd, and you're making a fool out of yourself.
At least be consistent with your argumentation. Above, you talked about the ability to "read the vni/pal file structure", now you're referring to the actual source code you've contributed.

On thing is clear, having a code active which already is outdated from the functional side could lead to more user unfriendly environment than removing VNI/PAL support and pointing to my DLL.

No, and that's why: Users who are already running with dmdext+VNI/PAL don't care the coloring is "outdated". If they would, they would already have switched to your DLL. Removing VNI/PAL from dmdext would break their setup. And users who are running your fork aren't impacted anyway.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

freezy commented:
Well, technically, it wasn't an insult. It's just a such ridiculous statement, given your history of violating the GPL and not caring about anyone's opinion in general.

But, anyway. :) 

Do you have a version of your DLL with the new API that works with both PAC and VNI?

lucky01 commented:
I sent you a new DLL which fixes the problem you had and should enable you to implement and test your interface with all coloring modes.

freezy commented:
Okay, I think I have it working. 

Would it be possible to add a `GetName(string name)` and `GetVersion(string version)` to the DLL, so I can print it and facilitate debugging? I'll add this to `DmdDevice.dll` as well, so PinballFX can show the version of the loaded DMD driver.

If you could add these methods to header file in DmdDevice's README that would be even better. :)

lucky01 commented:
I updated the testing dll on dropbox.

lucky01 commented:
Please make sure that all PinMame data is forwarded to the plugin even if no colorization is found and therefore no RGB24 data returned. I have implemented my own dump routine in the plugin to be able to dump fx3/tpa dmdext and alphanumeric data for colorization.

freezy commented:
Yup, that should be fine.

freezy commented:
@lucky01 quick question for me to understand: How do the `Advanced128x32` / `Advanced192x64` / `Advanced256x64` modes work? Does it mean that with `Advanced256x64`, a 128x32 source frame will be returned as 256x64 colored from the plugin? 

And how does it work with the global scaling options? The global scaling is done outside of the DLL, right?

lucky01 commented:
> Does it mean that with `Advanced256x64`, a 128x32 source frame will be returned as 256x64 colored from the plugin?

It simply describes that format the colorization project is in. So Advanced 256x64 is either FoorballFlipper which has 256x64 as source and will return 256x64 or a 128x32 source with HD frames as return. 128x32 scaling is done outside the plugin by scaling RGB24 data.

freezy commented:
Thanks, so Advanced 256x64 *always* returns 256x64, right? Do you have an example of a 128x32 game colorized at Advanced256x64 that I could test?

Also, just to confirm, the mode that you receive from `ColorizeGameSettings` won't change anymore during the game, right? So `width` and `height` could be set at initialization after receiving the mode, instead of during each `Convert()`?

freezy commented:
I'm also not quite getting this block:
```cs
if (frame.Data.Length == 128 * 32)
	Rgb24Buffer = Colorize4GrayWithRaw(128, 32, frame.Data, (ushort)vd.RawPlanes.Length, RawBuffer);
else if (frame.Data.Length == 192 * 64)
	Rgb24Buffer = Colorize4GrayWithRaw(192, 64, frame.Data, (ushort)vd.RawPlanes.Length, RawBuffer);
else if (frame.Data.Length == 256 * 64)
	Rgb24Buffer = Colorize4GrayWithRaw(256, 64, frame.Data, (ushort)vd.RawPlanes.Length, RawBuffer);
```
Doesn't this do the same?

```cs
Rgb24Buffer = Colorize4GrayWithRaw((ushort)frame.width, (ushort)frame.height, frame.Data, 
    (ushort)vd.RawPlanes.Length, RawBuffer);

```

lucky01 commented:
> Thanks, so Advanced 256x64 _always_ returns 256x64, right? Do you have an example of a 128x32 game colorized at Advanced256x64 that I could test?

I sent you a test project

> Also, just to confirm, the mode that you receive from `ColorizeGameSettings` won't change anymore during the game, right?

Right !

lucky01 commented:
> Wouldn't this do the same?

Only if frame.width and frame.height matches the size of the frame data, which is not the case if HDScaling is active.

freezy commented:
> Only if frame.width and frame.height matches the size of the frame data, which is not the case if HDScaling is active.

Yeah, I debugged into it, and indeed if HD scale is set, the frame size doesn't correspond to the data. I would consider this as a bug, and the scaling should be done after the colorization, if the colorizer doesn't already upscale it. Something like this:

![image](https://github.com/freezy/dmd-extensions/assets/70426/11549a5b-8984-4e20-be80-8476ccb36a3a)

Would that make sense?

lucky01 commented:
Either there are a lot of paths missing or wrong in your graph or I just don´t get it.
Anyway, scaling IS done after colorization already and the width/height parameters are just taken to indicate that scaling is needed. I´m ok how it is but if you consider this as a bug, feel free to change it.

freezy commented on LibDmd/Converter/Pin2Color/Pin2Color.cs:
```diff
@@ -0,0 +1,522 @@
+
+		private static bool _hasEvents { get; set; }
+
+		public enum ColorizerMode
+		{
+			None = -1,
+			SimplePalette = 0,
+			Advanced128x32 = 1,
+			Advanced192x64 = 3,
+			Advanced256x64 = 4,
+		}
```
I'm testing SAM games (ACDC), and I can't get it to work. I'm posting this here and not on my PR, since it doesn't seem work here, either. I think the problem is that there is no way of retrieving the palette changes from the plugin.

About the code above: Could you confirm (or correct me) that:
- `None` means that no colorization files were found, thus nothing will be colored. There will be a way to forward frames to the plugin nevertheless (for frame dumping), but nothing is expected in return.
- `SimplePalette` means that we'll only get palette changes from the plugin, for example for ACDC with a patched ROM.
- `Advanced128x32` means that all frames are colored and returned by the plugin, at 128x32.
- `Advanced192x64` also means that **all** frames come from the plugin, at **192x64**.
- `Advanced256x64` same as above, but at 256x64.

freezy commented:
Right, so ACDC with only a .pal should return `SimplePalette`, right? (It returns `None`.)

lucky01 commented:
It is not a simple palette but multiple palettes with ROM based palette switching. I don't remember what it reports back but I think it should be simple palette. Am 29.05.2023 um 14:35 schrieb freezy ***@***.***>:﻿
Right, so ACDC with only a .pal should return SimplePalette, right? (It returns None.)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

lucky01 commented:
You need to make sure the console data is forwarded to the pluginAm 29.05.2023 um 14:35 schrieb freezy ***@***.***>:﻿
Right, so ACDC with only a .pal should return SimplePalette, right? (It returns None.)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

freezy commented:
> It is not a simple palette but multiple palettes with ROM based palette switching.

Yes, so how about a `MultiPalette` mode, where dmdext can query the plugin for the palette with `GetCurrentPalette()` (returning an array of colors)?

> You need to make sure the console data is forwarded to the plugin

Yes, [that's the case](https://github.com/freezy/dmd-extensions/blob/plugins/LibDmd/Converter/Plugin/ColorizationPlugin.cs#L167).

lucky01 commented:
> Yes, so how about a `MultiPalette` mode, where dmdext can query the plugin for the palette with `GetCurrentPalette()` (returning an array of colors)?

I think that is not worth the time since there are only a hand full of smartdmd tables which will be converted anyway (e.g. Metallica)

freezy commented:
> I think that is not worth the time since there are only a hand full of smartdmd tables which will be converted anyway (e.g. Metallica)

I don't understand.. There are already colored games, how do you want to support them otherwise?

lucky01 commented:
> I don't understand.. There are already colored games, how do you want to support them otherwise?
I updated your testing DLL in dropbox. It was reporting the wrong colorization type.

freezy commented:
Cool, now how do I get the new palettes?

lucky01 commented:
You get the colored RGB24Am 30.05.2023 um 11:49 schrieb freezy ***@***.***>:﻿
Cool, now how do I get the new palettes?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

freezy commented:
That sucks. What you're proposing is:

- Your DLL converts a gray4-frame to RGB24 based on the current palette
- dmdext extracts the palette from the RGB24 frame
- dmdext creates a new colored frame with original frame data and the extracted palette

Instead of:

- Your DLL just returns the palette
- dmdext adds the palette to the frame

lucky01 commented:
That is what we agreed on . My plugin returns colorized RGB24 frames. Am 30.05.2023 um 12:46 schrieb freezy ***@***.***>:﻿
That sucks. What you're proposing is:

Your DLL converts a gray4-frame to RGB24 based on the current palette
dmdext extracts the palette from the RGB24 frame
dmdext creates a new colored frame with original frame data and the extracted palette

Instead of:

Your DLL just returns the palette
dmdext adds the palette to the frame


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

freezy commented:
Well, technically we agreed on not sending colored grayX-frames for pin2color-generated frames. Here, the palette changes are triggered by the ROM.

I'm trying to understand your motivation for such a shitty solution, because I can't see any benefits for anyone.

lucky01 commented:
Technically those ARE colored gray X (4) frames generated by the pin2color plugin and therefore returned as RGB24.All colored frames returned by the plugin are RGB24 . That is the consent we have and that is how it is implemented in my pull request.Am 30.05.2023 um 14:46 schrieb freezy ***@***.***>:﻿
Well, technically we agreed on not sending colored grayX-frames for pin2color-generated frames. Here, the palette changes are triggered by the ROM.
I'm trying to understand your motivation for such a shitty solution, because I can't see any benefits for anyone.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

lucky01 commented:
Intensive tests of my plugin solution have been made by the community and no drawback of any kind has been reported.
It seems all users of my plugin solution are happy.
The only thing I can agree with you is that we disagree on every point we discuss, Makes no sense to start the discussion again.

freezy commented:
> Intensive tests of my plugin solution have been made by the community 

But you've just changed the entire API. Nobody has ever tested what we're discussing here. This argument is nonsense.

> The only thing I can agree with you is that we disagree on every point we discuss, Makes no sense to start the discussion again.

Collaborator of the year ;)

Look, I'm not particularly enjoying this either. But I'm trying to put my feelings away so we can find a good solution. A good solution would be to retrieve the palette, so we can simply apply it to the frame and be done with it, instead of looping through the byte array multiple times.

lucky01 commented:
> But you've just changed the entire API. Nobody has ever tested what we're discussing here. This argument is nonsense.

Since may the 3rd the version which is for download on VPU and intensively tested is based on the latest code on my fork and IS using exactly the same API like you.

freezy commented:
> Since may the 3rd the version which is for download on VPU and intensively tested is based on the latest code on my fork and IS using exactly the same API like you.

That's what happens when using this PR with the pin2color.dll from VPU:

```
 [1] 2023/05/31 22:27:39.110  INFO | [Pin2Color] Successfully loaded colorizer plugin ... 
 [1] 2023/05/31 22:27:39.155  INFO | [Pin2Color] Failed to open colorizer ... 
```
Which makes sense, since yourself said:

> I sent you a dll version for testing. I refactored the coloring plugin interface to use the BSD-3 licensed dmddevice.dll api.

So, it's a different API, which is fine, it's just not the one the community tested.

Anyway, I'll be soon releasing a beta version to test, so more people can check for regressions. The pin2color.dll testing will be done when you're back from holidays, I guess.

About alphanumeric colorizations: Is there anything I could test with? Do you have one yourself that you could test?

freezy commented on LibDmd/Converter/Pin2Color/Pin2Color.cs:
```diff
@@ -0,0 +1,522 @@
+		private delegate IntPtr _dColorizeGetVersion();
+		private static _dColorizeGetVersion ColorizeGetVersion;
+
+		private static void processEvent()
+		{
+			if (_activePinUpOutput != null) {
+				uint eventID = ColorizeGetEvent();
+				if (eventID != lastEventID) {
+					lastEventID = eventID;
+					_activePinUpOutput.SendTriggerID((ushort)eventID);
+				}
+			}
+		}
```
AFAIK this is for an unreleased PinUp feature that was added to Serum recently. Did you just copy it blindly, or is there a way I can test this?
